[ROCm] Support lowering through PJRT_Triton_Extension

amd-jianli12 opened the pull request:
This change adds support for lowering Pallas Triton calls to HSACO through PJRT_Triton_Extension for ROCm.

As counterpart of [jax-ml/jax#25903](https://github.com/jax-ml/jax/pull/25903).

PR for XLA is at [openxla/xla#31284](https://github.com/openxla/xla/pull/31284)

superbobry reviewed (approved):
Thanks! 

JFYI this code pass is, unfortunately, not live at the moment. We attempted to switch to it a few times, but were blocked on internal regressions.

amd-jianli12 committed:
[ROCm] Support lowering through PJRT_Triton_Extension

This change adds support for lowering Pallas Triton calls to HSACO
through PJRT_Triton_Extension for ROCm.

dimitar-asenov commented:
I assume this PR should be merged after the XLA PR:
https://github.com/openxla/xla/pull/31284

The XLA PR makes changes to the `xla::triton::CompilationResult` struct that need to be reflected in JAX here:
https://github.com/jax-ml/jax/blob/main/jaxlib/gpu/gpu_plugin_extension.cc#L53-L78

@amd-jianli12 Please make the appropriate changes to that file in this PR. I would like to have a look at those before we merge the XLA PR.

dimitar-asenov reviewed (commented):
I had a quick look, but @superbobry should weigh in and review the new changes.

amd-jianli12 commented on jax/_src/lib/triton.py:
```diff
@@ -20,10 +20,8 @@
 
 class CompilationResult(Protocol):
   asm: str
+  hsaco_path: str
```
Yes, it's better to clarify and avoid confusion here. Will update.

amd-jianli12 commented on jax/_src/lib/triton.py:
```diff
@@ -19,11 +19,9 @@
 class CompilationResult(Protocol):
-  asm: str
+  asm: str  # Only set for CUDA (contains PTX)
+  hsaco_path: str  # Only set for ROCm (path to HSACO binary)
```
Yes, passing hsaco data is the solution. And it seems better to use this data vector or its hash rather than file path as key to module image cache. I can refine the implementation.

amd-jianli12 commented on jax/_src/pallas/triton/pallas_call_registration.py:
```diff
@@ -166,13 +169,15 @@ def pallas_call_lowering(
       compilation_result.smem_bytes,
-      compilation_result.asm,
+      (
+          compilation_result.hsaco_path
```
My understanding is each invocation should get a new file.
